File: S03/week06/sejun/searchInsertPosition.java
//Time Limit Exceeded

class Solution {
		public int searchInsert(int[] nums, int target) {
			int size = nums.length - 1;
			int left = 0;
			int mid = size / 2;
			int right = size;

			while (true) {
				if (nums[mid] == target) {
					return mid;
				}
				int temp = mid;

				if (target < nums[mid]) {
					right = mid;
					mid = ((mid - left) / 2);
				}

				if (nums[mid] < target) {
					left = mid;
					int nextIndex = ((right + mid) / 2);
					mid = mid + 1 > nextIndex ? mid + 1 : nextIndex;
				}

				if (mid >= size + 1 || temp == mid) {
					break;
				}
			}

			return mid;
		}
	}

File: S03/week06/seokjoon/SearchInsertPosition.kt
//178ms, 37.6mb
class Solution {
    fun searchInsert(nums: IntArray, target: Int): Int {
        var l = 0
        var r = nums.lastIndex
        var idx = (l + r) / 2

        while(l<=r){
            if(nums[idx] == target){
                return idx
            }else if(nums[idx] < target){
                l = idx + 1
            }else {
                r = idx - 1
            }
            idx = (l + r) / 2
        }
        return if(nums[idx] > target) 0 else idx + 1
    }
}
